More API guidelines

cole-h opened the pull request:
I'm not totally told on this, but I think it might be a good idea. I dunno.

---

Based on #3.

cole-h committed:
More API guidelines

cole-h committed:
Actual NixConfig struct, more docs, optional serde feature

cole-h committed:
wip: from string impls

cole-h committed:
wip: basically flattens to the contents of its single field

cole-h committed:
update readme

cole-h committed:
more into()

cole-h committed:
associated functions rather than freeform

cole-h committed:
fixup readme to reflect associated functions

cole-h committed:
fixup: use Self

cole-h committed:
update tempfile to drop vulnerable remove_dir_all version

cole-h commented:
I dunno, I kinda like this now...

cole-h committed:
bump tempfile in cargo.toml too

Hoverbear commented on src/lib.rs:
```diff
@@ -1,168 +1,236 @@
+/// value.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct NixConfigValue(pub String);
```
Is this always a string? Are booleans not a thing?

cole-h replied:
The underlying config file format in Nix is not typed, so I don't think it matters: https://github.com/NixOS/nix/blob/0079d2943702a7a7fbdd88c0f9a5ad677c334aa8/src/libutil/config.cc#L128

cole-h commented on src/lib.rs:
```diff
@@ -1,168 +1,236 @@
 #[derive(Clone, Eq, PartialEq, Debug, Default)]
-pub struct NixConfig(pub HashMap<String, String>);
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(transparent))]
```
I think this is the thing I'm most meh on... If we add new fields, we wouldn't be able to do this anymore (without implementing the `serde` stuff by hand. I could drop this attr, and the receipt over in `nix-installer` would then look like `nix_config: { settings: {  experimental-features = flakes, cores = 42 } }`, etc...

Hoverbear commented on src/lib.rs:
```diff
@@ -1,168 +1,236 @@
-        if let Some(pos) = line.find('#') {
-            line = &line[..pos];
-        }
+    pub fn to_settings(&self) -> HashMap<NixConfigKey, NixConfigValue> {
```
Since `NixConfig` already implements `Clone` and we offer `into_settings()` I'm not sure this is necessary API but that's fine.

cole-h committed:
drop newtype wrappers

they don't really add anything and don't help enforce any invariants so
wholly unnecessary

cole-h committed:
nesting is okay

cole-h commented:
OK I think I like this now. Let's do a 0.1.0!
